sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala

```diff
@@ -984,37 +1040,47 @@ case class GpuParquetMultiFilePartitionReaderFactory(
       files: Array[PartitionedFile],
       conf: Configuration): PartitionReader[ColumnarBatch] = {
     val clippedBlocks = ArrayBuffer[ParquetSingleDataBlockMeta]()
-    val currentTime = System.nanoTime()
-    files.map { file =>
-      val singleFileInfo = try {
-        filterHandler.filterBlocks(footerReadType, file, conf, filters, readDataSchema)
-      } catch {
-        case e: FileNotFoundException if ignoreMissingFiles =>
-          logWarning(s"Skipped missing file: ${file.filePath}", e)
-          ParquetFileInfoWithBlockMeta(new Path(new URI(file.filePath)), Seq.empty,
-            file.partitionValues, null, null, false, false, false)
-        // Throw FileNotFoundException even if `ignoreCorruptFiles` is true
-        case e: FileNotFoundException if !ignoreMissingFiles => throw e
-        // If ignoreMissingFiles=true, this case will never be reached. But it's ok
-        // to leave this branch here.
-        case e@(_: RuntimeException | _: IOException) if ignoreCorruptFiles =>
-          logWarning(
-            s"Skipped the rest of the content in the corrupted file: ${file.filePath}", e)
-          ParquetFileInfoWithBlockMeta(new Path(new URI(file.filePath)), Seq.empty,
-            file.partitionValues, null, null, false, false, false)
-      }
+    val startTime = System.nanoTime()
+
+    def addBlockMeta(metaAndFile: BlockMetaWithPartFile): Unit = {
+      val singleFileInfo = metaAndFile.meta
       clippedBlocks ++= singleFileInfo.blocks.map(block =>
         ParquetSingleDataBlockMeta(
           singleFileInfo.filePath,
           ParquetDataBlock(block),
-          file.partitionValues,
+          metaAndFile.file.partitionValues,
           ParquetSchemaWrapper(singleFileInfo.schema),
           singleFileInfo.readSchema,
           new ParquetExtraInfo(singleFileInfo.isCorrectedRebaseMode,
-            singleFileInfo.isCorrectedInt96RebaseMode, singleFileInfo.hasInt96Timestamps)))
+            singleFileInfo.isCorrectedInt96RebaseMode,
+            singleFileInfo.hasInt96Timestamps)))
+    }
+
+    if (numFilesFilterParallel > 0) {
+      val tc = TaskContext.get()
+      val tasks = new java.util.ArrayList[Future[Array[BlockMetaWithPartFile]]]()
+      logDebug("Using parallel threads to do the filter with the Coalescing reader, " +
+        s"$numFilesFilterParallel files per thread")
+      files.sliding(numFilesFilterParallel, numFilesFilterParallel).foreach { fileGroup =>
+        val threadPool = MultiFileReaderThreadPool.getOrCreateThreadPool(numThreads)
+        tasks.add(threadPool.submit(
+          new CoalescingFilterRunner(footerReadType, tc, fileGroup, conf, filters, readDataSchema)))
+      }
+      for (future <- tasks.asScala) {
+        val fileBlockMetaArray = future.get()
+        fileBlockMetaArray.foreach { metaAndFile =>
+          addBlockMeta(metaAndFile)
+        }
+      }
```

Comment:
Seems like this could be made more concise and eliminate the need for an explicitly managed array if we write it as something like:
```scala
  val threadpool = MultiFileReaderThreadPool.getOrCreateThreadPool(numThreads)
  files.grouped(numFilesFilterParallel).map { fileGroup =>
    threadPool.submit(
      new CoalescingFilterRunner(footerReaderType, tc, fileGroup, conf, filters, readDataSchema))
  }.toArray.foreach(_.get().foreach(addBlockMeta))
```
And then could put the body of addBlockMeta into the last foreach if desired, eliminating the local func.

Reply:
yes much cleaner, thanks.